File: 2593-find-score-of-an-array-after-marking-all-elements/2593-find-score-of-an-array-after-marking-all-elements.py
class Solution:
    def findScore(self, nums: List[int]) -> int:
        score = 0
        
        heap = []
        
        for i, num in enumerate(nums):
            heappush(heap, (num, i))
            
        visited = set()
        
        n = len(nums)
            
        for i, num in enumerate(nums):                
            if heap:
                get = heappop(heap)
                if get not in visited:
                    score += get[0]
                    visited.add((num, get[1]))
                    if get[1] - 1 >= 0:
                        visited.add((nums[get[1] - 1], get[1] - 1))
                    if get[1] + 1 < n:
                        visited.add((nums[get[1] + 1], get[1] + 1))
        
        return score

File: 2593-find-score-of-an-array-after-marking-all-elements/2593-find-score-of-an-array-after-marking-all-elements.java
import java.util.*;

class Solution {
    public long findScore(int[] nums) {
        final int n = nums.length;
        long score = 0;

        PriorityQueue<int[]> minHeap = new PriorityQueue<>((a, b) -> {
            if (a[0] == b[0]) {
                return Integer.compare(a[1], b[1]);
            }
            return Integer.compare(a[0], b[0]);
        });

        for (int i = 0; i < n; i++) {
            minHeap.offer(new int[]{nums[i], i});
        }

        Set<Integer> marked = new HashSet<>();

        while (!minHeap.isEmpty()) {
            int[] current = minHeap.poll();
            int value = current[0];
            int index = current[1];

            if (marked.contains(index)) {
                continue;
            }

            score += value;

            marked.add(index);
            
            if (index > 0) {
                marked.add(index - 1);
            }
            
            if (index < n - 1) {
                marked.add(index + 1);
            }
        }

        return score;
    }
}
